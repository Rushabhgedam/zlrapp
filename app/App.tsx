import { ApolloClient, ApolloProvider, InMemoryCache } from '@apollo/client';
import { NavigationContainer } from '@react-navigation/native';
import { createNativeStackNavigator } from '@react-navigation/native-stack';
import React from 'react';
import { StyleSheet } from 'react-native';
import HomeScreen from './homescreen';
import SplashScreen from './splashscreen';

const client = new ApolloClient({
  cache: new InMemoryCache(),
  uri: "https://countries.trevorblades.com/graphql"
})

const App = () => {
  const Stack = createNativeStackNavigator()

  return (
    <ApolloProvider client={client}>
      <NavigationContainer>
        <Stack.Navigator screenOptions={{
          headerShown: false
        }}>
          <Stack.Screen name='Splashscreen' component={SplashScreen} />
          <Stack.Screen name='homescreen' component={HomeScreen} />
        </Stack.Navigator>
      </NavigationContainer>

    </ApolloProvider>

  )
}

export default App

const styles = StyleSheet.create({})