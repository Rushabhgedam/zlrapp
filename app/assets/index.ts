export const selected = require("./selected.png");
export const unselected = require("./unselected.png");