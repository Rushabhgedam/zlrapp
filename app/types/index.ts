export interface IProps {
    navigation: { 
        navigate: (screenName: string) => void,
        replace: (screenName: string) => void 
    }
}


export interface IContinent {
    "name": string,
    "code": string,
    "selected"?: boolean
}


export interface ICountryItem {
    name: string,
    emoji: string,
    continent: { name: string }
}