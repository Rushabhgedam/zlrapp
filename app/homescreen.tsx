import { useQuery } from '@apollo/client'
import React, { useState } from 'react'
import { ActivityIndicator, Dimensions, FlatList, Image, RefreshControl, SafeAreaView, StatusBar, StyleSheet, Text, TextInput, TouchableOpacity, View } from 'react-native'
import { selected, unselected } from './assets'
import { continentsList } from './constants'
import { GET_COUNTRIES_OF_CONTINENTS } from './graphql/queries'
import { IContinent, ICountryItem, IProps } from './types/index'

const { width, height } = Dimensions.get("screen")

const HomeScreen = (props: IProps) => {
    const [continent, setContinent] = useState({
        code: "",
        name: ""
    })
    const [searchText, setSearchText] = useState('')

    const { loading, error, data, refetch, called } = useQuery(GET_COUNTRIES_OF_CONTINENTS, {
        variables: { code: continent.code },
    });

    const renderContinentsList = ({ item, index }: { item: IContinent, index: number }) => {
        let isSelected = item.code === continent.code
        return <TouchableOpacity testID={'continent' + index} onPress={() => setContinent({ code: item.code, name: item.name })}
            style={[styles.continentItem, { backgroundColor: isSelected ? "#e8f2fb" : "white" }]}>
            <Image source={isSelected ? selected : unselected} style={styles.radioBtn} />
            <Text style={styles.continentName}>
                {item.name}
            </Text>
        </TouchableOpacity>
    }

    const renderCountriesList = ({ item, index }: { item: ICountryItem, index: number }) => {
        return <View style={styles.row}>
            <Text style={styles.flag}>{item.emoji}</Text>
            <View>
                <Text style={styles.countryName}> {item.name}</Text>
                <Text style={styles.continentName1}> {item.continent.name}</Text>
            </View>
        </View>
    }
    const emptyCountry = () => <View>
        {
            continent.code != "" && loading ? <ActivityIndicator /> :
                searchText.length > 0 ? <Text>Not found results for {searchText}</Text> :
                    <Text>Select Continent</Text>
        }
    </View>
    const Seperator = () => <View style={{ marginTop: 10 }} />
    return (
        <SafeAreaView style={{flex:1}}>
        <View style={styles.container}>
            <StatusBar />
            <Seperator />
            <TextInput
                onChangeText={(t) => {
                    setSearchText(t)
                }}
                placeholder={continent.code === "" ? "Please select continent" : "Search Country here..."}
                style={styles.searchBar} />
            <Seperator />
            <View>
                <FlatList
                    data={continentsList}
                    keyExtractor={(_, index) => index.toString()}
                    ItemSeparatorComponent={() => <View style={{ paddingVertical: 2 }} />}
                    renderItem={renderContinentsList}
                />
            </View>
            <Seperator />
            <View>
                {continent.code && <Text style={{alignSelf:"center", margin:10}}>{`Available Countries in ${continent.name} are`}</Text>}
                <FlatList
                    ListEmptyComponent={emptyCountry}
                    data={data?.continent?.countries.filter((country: ICountryItem) => country.name.toLowerCase().includes(searchText.toLowerCase()))}
                    renderItem={renderCountriesList}
                    refreshControl={<RefreshControl
                        refreshing={loading}
                        onRefresh={refetch}
                    />}
                    contentContainerStyle={{ paddingBottom: 20 }}
                    keyExtractor={(_, index) => index.toString()}
                    style={styles.countriesFlatList}
                />
            </View>
        </View>
        </SafeAreaView>
    )
}

export default HomeScreen

const styles = StyleSheet.create({
    container:{ backgroundColor: "white", flex: 1 },
    continentItem: {
        flexDirection: "row",
        marginHorizontal: 6,
        paddingHorizontal: 6,
        borderRadius: 6,
    },
    radioBtn: {
        tintColor: "skyblue",
        height: 20,
        width: 20,
        resizeMode: "contain"
    },
    continentName: { color: "black" },
    continentName1: { color: '#b8b8b8' },
    row: { flexDirection: "row" },
    flag: { fontSize: 25, color: '#000' },
    countryName:{ color: '#000', fontSize: 16 },
    searchBar:{ width: "95%", borderRadius: 8, padding: 10, borderWidth: 0.5, alignSelf: "center" },
    countriesFlatList:{ marginBottom: height * 0.4, marginHorizontal: 6 }
})