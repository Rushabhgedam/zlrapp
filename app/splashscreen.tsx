import { ActivityIndicator, StyleSheet, Text, View } from 'react-native'
import React, { useEffect } from 'react'
import { IProps } from './types';


const SplashScreen = (props: IProps) => {
    useEffect(() => {
        setTimeout(() => {
            props.navigation.navigate("homescreen")
        }, 3000);
    }, [])
    return (
        <View style={{justifyContent:"center", alignItems:"center", flex:1}}>
            <Text>Welcome to Zeller</Text>
            <ActivityIndicator />
        </View>
    )
}

export default SplashScreen

const styles = StyleSheet.create({})