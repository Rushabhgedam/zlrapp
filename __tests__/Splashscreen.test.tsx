/**
 * @format
 */

import 'react-native';
import React from 'react';
import App from '../app/App';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import SplashScreen from '../app/splashscreen';
import { IProps } from '../app/types';
import { render, screen } from '@testing-library/react-native';
jest.useFakeTimers();

describe('Splash Screen test suite', () => {
  const navigationMock = {
    navigate: jest.fn(),
  };
  it('renders the initial component', () => {
    render(<SplashScreen navigation={navigationMock} />);
    expect(screen).toMatchSnapshot()
  });
  it('navigates to homescreen after 3 seconds', () => {
    jest.advanceTimersByTime(3000);
    expect(navigationMock.navigate).toHaveBeenCalledWith('homescreen');
  });
});