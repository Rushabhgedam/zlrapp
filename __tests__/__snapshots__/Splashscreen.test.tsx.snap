// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Splash Screen test suite renders the initial component 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text>
    Welcome to Zeller
  </Text>
  <ActivityIndicator />
</View>
`;
